### DISCUSSION TITLE
Is it possible to change the JS runtime used as the eval target?

### DISCUSSION THREAD

**Discussion by mjmeintjes** [thheller/shadow-cljs#1019] [Q&A] [answered]
In the [manual](https://shadow-cljs.github.io/docs/UsersGuide.html#_javascript_runtimes) it states that shadow-cljs "by default picks the first JS runtime that connected as the eval target."

Is there a way to change the eval target (without closing the first browser)? Or set it to use the last connected JS runtime by default?

**Comment by thheller** [comment, marked as answer]
Yes, this is possible. It is however somewhat manual currently.

First off you need to find the correct "runtime-id" you want to eval in. You can get that in a variety of ways:

- The runtime itself will list this on startup in the `shadow-cljs: #<an-integer> ready!` console message. With the integer being the runtime-id. This can be tricky to identify the correct one though if you use one or more workers. Since they all log the same message with different numbers. Otherwise it'll be just one and easy to see.
- The UI lists all runtimes and their ID. Defaults to http://localhost:9630/runtimes, again the #<an-integer>. It provides a little more info and differentiates between regular Browser and Workers. It also auto-updates when new ones arrive or old ones disappear.
- Manually pull the Data from the CLJ REPL. eg. `npx shadow-cljs clj-repl` and then `(shadow/repl-runtimes :your-build-id-here)`. It'll list the runtimes in EDN. You can do this via your Editor connection. Pretty printed it'll look something like this. For reasons this is called `:client-id` here though.
```clojure
[{:since #inst "2022-06-03T07:38:54.075-00:00",
  :proc-id "86abce83-e751-4eec-af91-4751d85fcb33",
  :connection-info {:remote true, :websocket true},
  :client-id 25,
  :user-agent "Chrome 537.36 [Win32]",
  :type :runtime,
  :lang :cljs,
  :build-id :browser,
  :host :browser,
  :worker-id 24,
  :dom true}
 {:since #inst "2022-06-03T07:38:54.369-00:00",
  :proc-id "86abce83-e751-4eec-af91-4751d85fcb33",
  :connection-info {:remote true, :websocket true},
  :client-id 26,
  :user-agent "Chrome 537.36 [Win32]",
  :type :runtime,
  :lang :cljs,
  :build-id :browser,
  :host :browser-worker,
  :worker-id 24,
  :dom false}]
```

Once you have the runtime-id you need you can currently only switch to it via the CLJ REPL.

```clojure
(shadow/repl :your-build-id-here {:runtime-id 26})
```

You can  `:cljs/quit` that REPL and switch to a different `:runtime-id` any time.

The default configuration picks the first Runtime that connects and stays with it while it is available. You can set `:repl {:runtime-select :latest}` in your `shadow-cljs.edn` top level or `~/.shadow-cljs/config.edn` to change this to always pick the last one that connected.
